sematic/resolvers/local_resolver.py

```diff
@@ -45,11 +47,78 @@ def __init__(self, **kwargs):
         self._buffer_runs: Dict[str, Run] = {}
         self._buffer_artifacts: Dict[str, Artifact] = {}
 
-        # TODO: Replace this with a local storage engine
-        self._store_artifacts = False
+        self._storage: Storage = LocalStorage()
 
         self._sio_client = socketio.Client()
 
+        self._rerun_from_run_id = rerun_from
+
+    def _seed_graph(self, future: AbstractFuture):
+        if self._rerun_from_run_id is None:
+            super()._seed_graph(future)
+        else:
+            self._seed_from_clone(future, self._rerun_from_run_id)
+
+    def _seed_from_clone(self, future: AbstractFuture, from_run_id: str):
+        try:
+            run = api_client.get_run(from_run_id)
+        except api_client.ResourceNotFoundError:
+            raise ValueError(f"Cannot restart from {from_run_id}: run cannot be found.")
+
+        if from_run_id == run.root_id:
+            raise ValueError(
+                f"Cannot restart from {from_run_id}: "
+                "this is the root run, simply start over"
+            )
+
+        runs, artifacts, edges = api_client.get_graph(run.root_id, root=True)
+
+        graph = Graph(runs=runs, artifacts=artifacts, edges=edges)
+
+        if not graph.input_artifacts_ready(run.id):
+            raise ValueError(
+                f"Cannot start from {from_run_id}: " "upstream runs did not succeed."
+            )
+
+        logger.info(f"Attempting to rerun from {from_run_id}")
+
+        cloned_futures, input_artifacts, output_artifacts = graph.clone_futures(
```

Comment:
We'll want to store in the `run` object that it was cloned and point to which run it was cloned from. We'll want to surface that in the UI. That can wait for a separate PR though.

Reply:
Yes, it was my intention to do a follow-up PR.